### ISSUE TITLE
loadRemoteModule with a remote of a remote

### ISSUE THREAD

**Issue by asollberger** [angular-architects/module-federation-plugin#379]
I'm trying to set up a set of repos where I'm loading routes in a chain:

The shell loads routes from the request repo and that one in turn should load routes from the request-parties repo.
The issue I'm having is that the remote of the middle repo is empty.
I tried adding the third remote to the host, add it to both, add it explicitly in the shared object, none of it worked.

The error I'm getting is:
```js
async function loadRemoteModule(optionsOrRemoteName, exposedModule) {
    const options = normalizeOptions(optionsOrRemoteName, exposedModule);
    await ensureRemoteInitialized(options);
    const remoteName = getRemoteNameByOptions(options);
    const remote = getRemote(remoteName);
    if (!remote) {
        throw new Error('unknown remote ' + remoteName);
    }
```

Is there a way to do this? (I'm currently working through your code to see whether I can inject it somewhere for the moment)

The **shell repo (host)** has this manifest:
```json
{ "request": "http://localhost:4210/remoteEntry.json" }
```
and these routes:
```js
export const routes: Routes = [
  { path: '', pathMatch: 'full', redirectTo: 'request-routes' },
  { path: 'request-component', loadComponent: () => loadRemoteModule('request', './RequestComponent').then((m) => m.AppComponent) },
  { path: 'request-routes', loadChildren: () => loadRemoteModule('request', './request-routes').then((m) => m.requestRoutes) }
];
```

The **request repo (remote)** has this manifest:
```json
{ "@cob/request-parties": "http://localhost:4220/remoteEntry.json" }
```
and these routes
```js
export const requestRoutes: Routes = [
  {path: '', pathMatch: 'full', redirectTo: 'request-parties-component'},
  {path: 'request', component: RequestComponent},
  {path: 'parties', component: RequestComponent, loadChildren: () => loadRemoteModule('@cob/request-parties', './request-parties-routes').then((m) => m.requestPartyRoutes)},
  {path: 'party', loadComponent: () => loadRemoteModule('@cob/request-parties', './Component').then((m) => m.AppComponent)},
  {path: 'request-parties-component', loadComponent: () => loadRemoteModule('@cob/request-parties', './Component').then((m) => m.AppComponent)  },
];
```
and this federation config:
```js
module.exports = withNativeFederation({
  name: 'request',
  exposes: {
    './RequestComponent': './projects/request/src/app/request.component.ts',
    './request-routes': './projects/request/src/app/request.routes.ts'
  },
  ...
}
```

and the request parties repo (another remote) has these routes:
```js
export const requestPartyRoutes: Routes = [
  {path: '', pathMatch: 'full', redirectTo: 'party'},
  {path: 'party', component: AppComponent}
];
```
and this federation config:
```js
module.exports = withNativeFederation({
  name: '@cob/request-parties',
  exposes: {
    './Component': './modules/request-parties/src/app/app.component.ts',
    './request-parties-routes': './modules/request-parties/src/app/app.routes.ts',
  },
   ...
}
```

**Comment by manfredsteyer** [comment]
Hi,

thanks for this message.

Can you please try to reference all the remote (also the indirect ones) in the outer shell's manifest?

**Comment by asollberger** [comment]
What Jean-Baptiste did is pretty much also what I did and my conclusions were approximately the same. I did try to publish twice though and that didn't work either. I'll post a reproduction of that shortly

**Comment by JBBianchi** [comment]
@asollberger if you're using dynamic hosts, a dirty trick would be to call `initFederation('/assets/federation.manifest.json')` where you export your 'request app' routes.

If [hasRemote](https://github.com/angular-architects/module-federation-plugin/blob/main/libs/native-federation-runtime/src/lib/model/remotes.ts#L27C17-L27C26) was exported, it would allow a slightly less messy approach:
```typescript
if (!hasRemote('<known remote>')) { // <-- e.g. 'parties'
    initFederation('/assets/federation.manifest.json');
}
```

**Comment by asollberger** [comment]
Here's my version of what I'm trying to achieve:
https://github.com/asollberger/nested-native-federation

I tried
- double manifests, didn't work because it tried to load the same dependencies again (shared ones like @angular/core)
- tried a reinit (initFederation) of the same one (same problem as above)

@JBBianchi I tried your suggestion. It did load another importmap-shim and added all the dependcies in there again, which caused duplicate errors.

For the moment I'm solving my immediate issue through this construct:
```js
export const shellRoutes: Routes = [
  {path: '', pathMatch: 'full', redirectTo: 'request/8/party'},
  {path: 'request-component', loadComponent: () => loadRemoteModule('request', './RequestComponent').then((m) => m.RequestComponent)  },
  {
    path: 'request/:id',
    loadComponent: () => loadRemoteModule('request', './RequestComponent').then((m) => m.RequestComponent),
    children: [
      {path: 'party', loadChildren: () => loadRemoteModule('@cob/request-parties', './request-parties-routes').then((m) => m.requestPartyRoutes)}
    ]
  },
    { path: 'request-router', loadChildren: () => loadRemoteModule('request', './request-routes').then((m) => m.requestRoutes)}
]
```
... but that's just obviously not solving the issue

**Comment by JBBianchi** [comment]
@asollberger that's unfortunate. I saw the duplicate errors in my console but it didn't seem to have an impact and it seemed to work as intended in the end.

What about something along the lines of the following?
```typescript
await loadManifest('/assets/federation.manifest.json')
  .then(processRemoteInfos);
```

It should populate `remoteNamesToRemote` without messing with the importmap.

**Comment by mchlbrnd** [comment]
I have following situation:

shell (dynamic host) -> remote-1 (remote/dynamic host) -> remote-2 (remote)

I resolved your issue by having  remote-1 load global manifest before calling loadRemoteModule. This does require to have all remotes defined in single manifest.

See https://github.com/wrckt/native-federation-decorators/blob/pre-esm-poc/apps/remote-1/src/app/app.routes.ts

This allows loading nested remote routes via shell (ie. http://shell/remote-1/remote-2)

Hope this helps!

**Comment by manfredsteyer** [comment]
Awesome. Yes, this is the current way of solving this. Thanks for providing an example.

**Comment by asollberger** [comment]
Thank you all for your comments, I truly appreciate all the feedback and help.

@mchlbrnd:
I do agree with @JBBianchi, the initFederation does too much and leaves errors in the console (though it does seem to work):
... for development that might be acceptable for a while, but not for production
```js
Error: Rejected map override "@angular/animations" from http://localhost:4200/_angular_animations-17_0_3-dev.js to ./_angular_animations-17_0_3-dev.js.
    at resolveAndComposePackages (es-module-shims.js:248:15)
    at resolveAndComposeImportMap (es-module-shims.js:201:7)
    at es-module-shims.js:925:23
```

The solution @JBBianchi provided with processRemoteInfo so far looks like the cleanest solution and does not have the override issue, thanks @JBBianchi.

It even works when splitting the manifest into pieces, the only issue with that (which kind of defeats the purpose of splitting it) is that you have to define the middleman's manifest in the shell. A way around that would be to explicitely load the manifest from the remote itself and for that I'd need to retrieve the remote's url:
while this doesn't find the manifest file:
`https://localhost:4200/assets/navigation.federation.manigest.json`
this call would find it:
`https://localhost:4202/assets/navigation.federation.manigest.json`

Another issue I noticed with all of the solution so far is the angular component id collision issue:
```js
_angular_core-17_0_3-dev.js:1157 NG0912: Component ID generation collision detected. Components '_ClientNavigationComponent' and '_ClientNavigationComponent' with selector 'nav-client-navigation' generated the same component ID. To fix this, you can change the selector of one of those components or add an extra host attribute to force a different ID. Find more at https://angular.io/errors/NG0912_angular_core-17_0_3-dev.js:1157 NG0912: Component ID generation collision detected. Components '_ClientNavigationComponent' and '_ClientNavigationComponent' with selector 'nav-client-navigation' generated the same component ID. To fix this, you can change the selector of one of those components or add an extra host attribute to force a different ID. Find more at https://angular.io/errors/NG0912
```
It happens when the same remote is attached in different places:
- calling the account directly from the shell
- calling the account through the navigation remote
I think this scenario will need some investigation, though it's not an issue for me at the moment.

I tried to initialize and pass a host attribute, initialized it before the call to the AppComponent and still got the warning.

**Comment by manfredsteyer** [comment]
`initFederation` should indeed only be called once. That means the remotes you load should not call it.

```
Shell:
[main.ts] --> [AppComponent]
                                  | Federation
MFE 1:                       V
[main.ts] --> [AppComponent]
                                  | Federation
MFE 2:                       V
[main.ts] --> [AppComponent]
```

Here, main.ts would call `initFederation`, hence it's only called once.

And if MFE 2 is mentioned in the Shell's manifest, it should work.

**Comment by gigadie** [comment]
My case might be slightly different, as I was only trying to render a component from "remote-2" inside "remote-1" other than in the "host/shell". I managed to get this to work simply by doing this in "remote-1":
```js
// my-component.html
[...]
<ng-container #remote2Component></ng-container>
[...]

// my-component.ts
@Component({
  [...]
})
export class MyComponent implements OnInit, AfterViewInit {

  @ViewChild('remote2Component', { read: ViewContainerRef, static: true }) viewContainer: ViewContainerRef | undefined;

  ngOnInit() {
    // [...]
  }

  async ngAfterViewInit(): Promise<void> {
    this.viewContainer?.clear();

    // this will load shell manifest, which in my case is totally fine
    // otherwise this can be configured to load local manifest, but it will need
    // full url "https://remote-1-url/assets/manifest.json
    await loadManifest('/assets/manifest.json').catch((err) => console.error(err));
    const m = await loadRemoteModule({
      type: 'manifest',
      remoteName: 'remote-2',
      exposedModule: './Remote2Component'
    });

    const instance = this.viewContainer?.createComponent(m.Remote2Component);
    // can interact with instance
    // console.log(instance);
  }
}
```

I guess a very similar approach can be used for routes.

**Comment by asollberger** [comment]
Just in case anyone stumbles across this and wants to do the same, here's a small fix of @JBBianchi's code that truly loads the manifest only once:
```js
let manifestLoaded : Subject<boolean> = null;

export async function fetchManifest(): Promise<void> {
    manifestLoaded = new ReplaySubject<boolean>();
    await fetch('/assets/federation.manifest.json')
        .then((manifest) => manifest.json())
        .then((manifestRemotes: Record<string, string>) =>
            Promise.allSettled(
                Object.entries(manifestRemotes).map(([federationRemoteName, federationInfoUrl]) => processRemoteInfo(federationInfoUrl, federationRemoteName))
            )
        )
        .then(() => manifestLoaded.next(true));
}

export async function loadRemoteModuleAfterManifest<T = any>(remoteName: string, exposedModule: string): Promise<T> {
    await (manifestLoaded ? firstValueFrom(manifestLoaded) : fetchManifest());
    return await loadRemoteModule(remoteName, exposedModule);
}

export const routes: Routes = [
    { path: '', pathMatch: 'full', redirectTo: 'accounts'},
    { path: 'accounts', component: AccountNavigationComponent, loadChildren: () => loadRemoteModuleAfterManifest('accounts', './routes').then((m) => m.routes) },
    { path: 'clients', component: ClientNavigationComponent, loadChildren: () => loadRemoteModuleAfterManifest('clients', './routes').then((m) => m.routes) }
];
```
